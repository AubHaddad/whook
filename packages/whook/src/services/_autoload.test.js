import initAutoload from './_autoload';
import { service } from 'knifecycle';

describe('$autoload', () => {
  const log = jest.fn();
  const require = jest.fn();

  beforeEach(() => {
    log.mockReset();
    require.mockReset();
  });

  it('should work for a config constant', async () => {
    require.mockReturnValueOnce({
      default: {
        CONFIG: {
          testConfig: 'test',
        },
      },
    });
    require.mockReturnValueOnce({
      default: service(async () => ({ info: {} }), 'API'),
    });

    const $autoload = await initAutoload({
      NODE_ENV: 'development',
      PWD: '/home/whoami/my-whook-project',
      SERVICE_NAME_MAP: {},
      INITIALIZER_PATH_MAP: {},
      WRAPPERS: [],
      require,
      log,
    });
    const result = await $autoload('CONFIG');

    expect({
      result,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      requireCalls: require.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work for API', async () => {
    require.mockReturnValueOnce({
      default: {
        CONFIG: {
          testConfig: 'test',
        },
      },
    });
    require.mockReturnValueOnce({
      default: service(async () => ({ info: {} }), 'API'),
    });

    const $autoload = await initAutoload({
      NODE_ENV: 'development',
      PWD: '/home/whoami/my-whook-project',
      SERVICE_NAME_MAP: {},
      INITIALIZER_PATH_MAP: {},
      WRAPPERS: [],
      require,
      log,
    });
    const result = await $autoload('API');

    expect({
      result,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      requireCalls: require.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work for handlers hash', async () => {
    require.mockReturnValueOnce({
      default: {
        CONFIG: {
          testConfig: 'test',
        },
      },
    });
    require.mockReturnValueOnce({
      default: service(
        async () => ({
          host: 'localhost:1337',
          swagger: '2.0',
          info: {
            version: '1.0.0',
            title: 'Sample Swagger',
            description: 'A sample Swagger file for testing purpose.',
          },
          basePath: '/v1',
          schemes: ['http'],
          paths: {
            '/ping': {
              get: {
                operationId: 'getPing',
                summary: "Checks API's availability.",
                responses: {
                  '200': {
                    description: 'Pong',
                  },
                },
              },
            },
          },
        }),
        'API',
      ),
    });
    require.mockReturnValueOnce({
      default: service(async () => async () => ({ status: 200 }), 'getPing'),
    });

    const $autoload = await initAutoload({
      NODE_ENV: 'development',
      PWD: '/home/whoami/my-whook-project',
      SERVICE_NAME_MAP: {},
      INITIALIZER_PATH_MAP: {},
      WRAPPERS: [],
      require,
      log,
    });
    const result = await $autoload('HANDLERS');

    expect({
      result,
      HANDLERS: await result.initializer({ getPing: () => {} }),
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      requireCalls: require.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work for handlers', async () => {
    require.mockReturnValueOnce({
      default: {
        CONFIG: {
          testConfig: 'test',
        },
      },
    });
    require.mockReturnValueOnce({
      default: service(async () => ({ info: {} }), 'API'),
    });
    require.mockReturnValueOnce({
      default: service(async () => async () => ({ status: 200 }), 'getPing'),
    });

    const $autoload = await initAutoload({
      NODE_ENV: 'development',
      PWD: '/home/whoami/my-whook-project',
      SERVICE_NAME_MAP: {},
      INITIALIZER_PATH_MAP: {},
      WRAPPERS: [],
      require,
      log,
    });
    const result = await $autoload('getPing');

    expect({
      result,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      requireCalls: require.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work with no wrappers', async () => {
    require.mockImplementationOnce(() => ({
      default: {
        CONFIG: {
          testConfig: 'test',
        },
      },
    }));
    require.mockImplementationOnce(() => {
      throw new Error('E_ERROR');
    });
    require.mockImplementationOnce(() => ({
      default: service(async () => ({ info: {} }), 'API'),
    }));
    require.mockImplementationOnce(() => ({
      default: service(async () => async () => ({ status: 200 }), 'getPing'),
    }));

    const $autoload = await initAutoload({
      NODE_ENV: 'development',
      PWD: '/home/whoami/my-whook-project',
      SERVICE_NAME_MAP: {},
      INITIALIZER_PATH_MAP: {},
      require,
      log,
    });
    const result = await $autoload('getPing');

    expect({
      result,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      requireCalls: require.mock.calls,
    }).toMatchSnapshot();
  });

  it('should work for wrapped handlers', async () => {
    require.mockReturnValueOnce({
      default: {
        CONFIG: {
          testConfig: 'test',
        },
      },
    });
    require.mockReturnValueOnce({
      default: service(
        async () => [async initHandler => initHandler],
        'WRAPPERS',
      ),
    });
    require.mockReturnValueOnce({
      default: service(async () => ({ info: {} }), 'API'),
    });
    require.mockReturnValueOnce({
      default: service(async () => async () => ({ status: 200 }), 'getPing'),
    });

    const $autoload = await initAutoload({
      NODE_ENV: 'development',
      PWD: '/home/whoami/my-whook-project',
      SERVICE_NAME_MAP: {},
      INITIALIZER_PATH_MAP: {},
      require,
      log,
    });
    const result = await $autoload('getPing');

    expect({
      result,
      logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
      requireCalls: require.mock.calls,
    }).toMatchSnapshot();
  });
});
