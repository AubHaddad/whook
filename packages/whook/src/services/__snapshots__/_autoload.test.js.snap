// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$autoload should work for API 1`] = `
Object {
  "logCalls": Array [
    Array [
      "debug",
      "ü§ñ - Initializing the \`$autoload\` service.",
    ],
    Array [
      "warning",
      "‚ö°Ô∏è - Loading configuration from /home/whoami/my-whook-project/config/development/config.",
    ],
    Array [
      "debug",
      "üíø - Loading API initializer from /home/whoami/my-whook-project/src/services/API.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/config/development/config",
    ],
    Array [
      "/home/whoami/my-whook-project/src/services/API",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "API",
    "path": "internal://API",
  },
}
`;

exports[`$autoload should work for a config constant 1`] = `
Object {
  "logCalls": Array [
    Array [
      "debug",
      "ü§ñ - Initializing the \`$autoload\` service.",
    ],
    Array [
      "warning",
      "‚ö°Ô∏è - Loading configuration from /home/whoami/my-whook-project/config/development/config.",
    ],
    Array [
      "debug",
      "üíø - Loading API initializer from /home/whoami/my-whook-project/src/services/API.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/config/development/config",
    ],
    Array [
      "/home/whoami/my-whook-project/src/services/API",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "CONFIG",
    "path": "internal://CONFIG",
  },
}
`;

exports[`$autoload should work for handlers 1`] = `
Object {
  "logCalls": Array [
    Array [
      "debug",
      "ü§ñ - Initializing the \`$autoload\` service.",
    ],
    Array [
      "warning",
      "‚ö°Ô∏è - Loading configuration from /home/whoami/my-whook-project/config/development/config.",
    ],
    Array [
      "debug",
      "üíø - Loading API initializer from /home/whoami/my-whook-project/src/services/API.",
    ],
    Array [
      "debug",
      "üíø - Loading getPing initializer from /home/whoami/my-whook-project/src/handlers/getPing.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/config/development/config",
    ],
    Array [
      "/home/whoami/my-whook-project/src/services/API",
    ],
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing",
  },
}
`;

exports[`$autoload should work for handlers hash 1`] = `
Object {
  "HANDLERS": Object {
    "getPing": [Function],
  },
  "logCalls": Array [
    Array [
      "debug",
      "ü§ñ - Initializing the \`$autoload\` service.",
    ],
    Array [
      "warning",
      "‚ö°Ô∏è - Loading configuration from /home/whoami/my-whook-project/config/development/config.",
    ],
    Array [
      "debug",
      "üíø - Loading API initializer from /home/whoami/my-whook-project/src/services/API.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/config/development/config",
    ],
    Array [
      "/home/whoami/my-whook-project/src/services/API",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "HANDLERS",
    "path": "internal://HANDLERS",
  },
}
`;

exports[`$autoload should work for wrapped handlers 1`] = `
Object {
  "logCalls": Array [
    Array [
      "debug",
      "ü§ñ - Initializing the \`$autoload\` service.",
    ],
    Array [
      "warning",
      "‚ö°Ô∏è - Loading configuration from /home/whoami/my-whook-project/config/development/config.",
    ],
    Array [
      "debug",
      "üíø - Loading WRAPPERS initializer from /home/whoami/my-whook-project/src/services/WRAPPERS.",
    ],
    Array [
      "debug",
      "üíø - Loading API initializer from /home/whoami/my-whook-project/src/services/API.",
    ],
    Array [
      "debug",
      "üíø - Loading getPing initializer from /home/whoami/my-whook-project/src/handlers/getPing.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/config/development/config",
    ],
    Array [
      "/home/whoami/my-whook-project/src/services/WRAPPERS",
    ],
    Array [
      "/home/whoami/my-whook-project/src/services/API",
    ],
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing",
  },
}
`;

exports[`$autoload should work with no wrappers 1`] = `
Object {
  "logCalls": Array [
    Array [
      "debug",
      "ü§ñ - Initializing the \`$autoload\` service.",
    ],
    Array [
      "warning",
      "‚ö°Ô∏è - Loading configuration from /home/whoami/my-whook-project/config/development/config.",
    ],
    Array [
      "debug",
      "üö´ - Could not find any API wrapper...",
    ],
    Array [
      "debug",
      "üíø - Loading API initializer from /home/whoami/my-whook-project/src/services/API.",
    ],
    Array [
      "debug",
      "üíø - Loading getPing initializer from /home/whoami/my-whook-project/src/handlers/getPing.",
    ],
  ],
  "requireCalls": Array [
    Array [
      "/home/whoami/my-whook-project/config/development/config",
    ],
    Array [
      "/home/whoami/my-whook-project/src/services/WRAPPERS",
    ],
    Array [
      "/home/whoami/my-whook-project/src/services/API",
    ],
    Array [
      "/home/whoami/my-whook-project/src/handlers/getPing",
    ],
  ],
  "result": Object {
    "initializer": [Function],
    "name": "getPing",
    "path": "/home/whoami/my-whook-project/src/handlers/getPing",
  },
}
`;
